### PR TITLE
Fix for late-joiners KEEP_LAST, RELIABLE, TRANSIENT_LOCAL [8986]

### PR DESCRIPTION
**Pull request by laura-eprosima**
* New large data volatile blackbox tests
* New large data transient local blackbox tests
* New blackbox tests to check #8896 bug (Transient Local, Reliable, Keep Last 1)
* Change the Subscriber History condition to check if the History is Full (Fix for the #8896 bug)

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2423/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2226/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2412/)

File: test/blackbox/common/BlackboxTestsPubSubHistory.cpp [hunk 1]
```diff
@@ -563,6 +563,110 @@ TEST(BlackBox, PubSubAsReliableKeepLastReaderSmallDepthWithKey)
...
+
+    ASSERT_FALSE(expected_data.empty());
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(500));
+    reader.startReception(expected_data);
+    eprosima::fastrtps::rtps::SequenceNumber_t seq;
+    seq.low = 10;
+    reader.block_for_seq(seq);
+
+    ASSERT_EQ(reader.get_last_sequence_received().low, 10);
```
Comment (MiguelCompany) [inline test/blackbox/common/BlackboxTestsPubSubHistory.cpp:621]:
```suggestion
    ASSERT_EQ(reader.get_last_sequence_received().low, 10u);
```

File: test/blackbox/common/BlackboxTestsPubSubHistory.cpp [hunk 2]
```diff
@@ -563,6 +563,110 @@ TEST(BlackBox, PubSubAsReliableKeepLastReaderSmallDepthWithKey)
...
+    // Wait for discovery.
+    writer.wait_discovery();
+    reader.wait_discovery();
+
+    reader.startReception(reader_data);
+    eprosima::fastrtps::rtps::SequenceNumber_t seq;
+    seq.low = 10;
+    reader.block_for_seq(seq);
+
+    ASSERT_EQ(reader.get_last_sequence_received().low, 10);
```
Comment (MiguelCompany) [inline test/blackbox/common/BlackboxTestsPubSubHistory.cpp:666]:
```suggestion
    ASSERT_EQ(reader.get_last_sequence_received().low, 10u);
```

File: src/cpp/fastrtps_deprecated/subscriber/SubscriberHistory.cpp [hunk 3]
```diff
@@ -57,17 +57,17 @@ SubscriberHistory::SubscriberHistory(
...
-                        topic_att.getTopicKind() == NO_KEY ?
-                            std::min(topic_att.resourceLimitsQos.allocated_samples, topic_att.historyQos.depth) :
-                            std::min(topic_att.resourceLimitsQos.allocated_samples, topic_att.historyQos.depth
-                                     * topic_att.resourceLimitsQos.max_instances),
-                topic_att.historyQos.kind == KEEP_ALL_HISTORY_QOS ?
-                        topic_att.resourceLimitsQos.max_samples :
-                        topic_att.getTopicKind() == NO_KEY ?
-                            topic_att.historyQos.depth :
-                            topic_att.historyQos.depth * topic_att.resourceLimitsQos.max_instances))
+            topic_att.historyQos.kind == KEEP_ALL_HISTORY_QOS ?
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastrtps_deprecated/subscriber/SubscriberHistory.cpp:61]: I think I was not clear enough. I was trying to tell you to separate the `HistoryAttributes` calculation to a separate static method.

[the remark this answers is not in the document]
Comment (laura-eprosima) [reply src/cpp/fastrtps_deprecated/subscriber/SubscriberHistory.cpp]: I had not understood it correctly, but now I got your point.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2425/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2228/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2414/)

**Comment by MiguelCompany** [comment]
Unrelated failures. Going in!

**Pull request merged**